dartagnan/src/main/java/com/dat3m/dartagnan/analysis/graphRefinement/graphs/ExecutionGraph.java

```diff
@@ -0,0 +1,318 @@
+    // We might want to add similar features for other linear orders (i.e. user defined orders)
+    // We also assume, that the non-transitive write order is defined.
+    public boolean addCoherenceEdges(Edge coEdge) {
+        return addCoherenceEdges(ImmutableList.of(coEdge));
+    }
+
+    public boolean addCoherenceEdges(Collection<Edge> coEdges) {
+        if ( woGraph == null) {
+            return false;
+        }
+        graphHierarchy.addEdgesAndPropagate(woGraph, coEdges);
+        return true;
+    }
+
+    public void backtrack() {
+        graphHierarchy.backtrack();
+    }
+
+    // =======================================================
+
+
+    //=================== Reading the WMM ====================
+
+    private Constraint getConstraintFromAxiom(Axiom axiom) {
+        if (constraintMap.containsKey(axiom)) {
+            return constraintMap.get(axiom);
+        }
+
+        Constraint constraint;
+        EventGraph innerGraph = getGraphFromRelation(axiom.getRelation());
+        if (axiom.isAcyclicity()) {
+            constraint = new AcyclicityConstraint(innerGraph);
+        } else if (axiom.isEmptiness()) {
+            constraint = new EmptinessConstraint(innerGraph);
+        } else if (axiom.isIrreflexivity()) {
+            constraint = new IrreflexivityConstraint(innerGraph);
+        } else {
+            throw new UnsupportedOperationException("The axiom " + axiom + " is not recognized.");
+        }
+
+        constraintMap.put(axiom, constraint);
+        return constraint;
+    }
+
+    private EventGraph getGraphFromRelation(Relation rel) {
+        if (relationGraphMap.containsKey(rel)) {
+            return relationGraphMap.get(rel);
+        }
+
+        EventGraph graph;
+        Class<?> relClass = rel.getClass();
+
+        // ===== Filter special relations ======
+        if (SPECIAL_RELS.contains(rel.getName())) {
+            switch (rel.getName()) {
+                case CTRL:
+                    graph = new CtrlDepGraph();
+                    break;
+                case DATA:
+                    graph = new DataDepGraph();
+                    break;
+                case ADDR:
+                    graph = new AddrDepGraph();
+                    break;
+                case CRIT:
+                    graph = new RcuGraph();
+                    break;
+                default:
+                    throw new UnsupportedOperationException(rel.getName() + " is marked as special relation but has associated graph.");
+            }
+        } else if (relClass == RelRf.class) {
+            graph = rfGraph = new ReadFromGraph();
+        } else if (relClass == RelLoc.class) {
+            graph = locGraph = new LocationGraph();
+        } else if (relClass == RelPo.class) {
+            graph = poGraph = new ProgramOrderGraph();
+        } else if (relClass == RelCo.class) {
+            graph = coGraph = new TransitiveGraph(woGraph = new CoherenceGraph());
```

Comment:
I never knew what to call the non-transitive version of coherence. If you are fine with "write order", then I will change it. But I'm also open to any other suggestion.

Reply:
The only think I "don't like" about "wo" and "write order" is that one might confuse it with the relation used by Viktor and his group which is basically "co". What about using "iwo" and "immediate write order"?

Reply:
"iwo" sounds good. But I guess it could also be confused with a co-like relation that only relates events that are adjacent in the co-order (i.e. iwo = co \ co;co).

Reply:
I though that was exactly what the `woGraph` was about, i.e. the minimal set of edges from which you can construct the full coherence.

Reply:
It is not minimal in general, it may contain edges that are transitively derivable from the other ones. I'm not sure how easy it is to maintain a minimal one, because lets say I add edges (w1, w2) and (w1, w3). If I add (w2, w3) now, then (w1, w2) becomes obsolete, similarly, if I add (w3, w2) then (w1, w2) will become obsolete. So I would need to delete either of the previously added ones. Generally, maintaining a minimal transitive reduction is as complex maintaining a transitive closure. That's why ```wo``` is just any relation with ```wo+ = co``` and not necessarily the smallest one.

Reply:
What about "swo" and "small write order"? I still want to keep write rather than coherence to avoid relating them as much as possible.

Reply:
I'm not a big fan of that naming. Maybe "reduced write order" or "reduced coherence"? But that might relate to "transitive reduction" which in turn is usually defined as a minimal graph with identical reachability properties. But I guess we cannot and should not be too specific with the naming (e.g. "NotMinimallyReducedCoherenceGraph" is worse than "ReducedCoherenceGraph").

Reply:
I also thought about "reduced" but I found it misleading for exactly what you mention.

I think the most correct term would be "minimal":
>Minimum: the least or smallest amount or quantity possible, attainable, or required. Minimal is a little more flexible, where it refers to being the smallest amount or degree in non-absolute terms

Reply:
But it is neither minimal nor is it a minimum graph. In fact, for transitive reductions on acyclic directed graphs both notions coincide as there is a unique minimal subgraph with the same reachability properites (i.e. there is a unique minimal transitive reduction which then also is the minimum by uniqueness.). And the current ```wo``` is not minimal by any definition. That's why I like words like ```reduced``` because they don't have to mean ```completely reduced``` or ```reduced to a minimum / maximally reduced```, but unfortunately, in the term "transitive reduction" it does refer to maximal reduction.

Reply:
Mathematically speaking, you are right. However, from a gramatical point of view, my understanding it that minimal implies SOME notion of minimum (in our case the minimum graph we can easily compute without having to remove edges), but it is flexible and it does not necessary implies smallest (I see it more related to small) in absolute terms.

Along the line of your last comment, another option could be `simplified`.

I'm fine with any of all options we mentioned.

Reply:
Since we are working in a mathematical environment, I wouldn't want to use minimum/minimal lightly. And to make the mathematical "minimal" fitting, we would need to stretch the definition of what "minimality" means by a lot, especially because it is possible to compute a smaller on relation. It is maybe "minimal" w.r.t. to effort, i.e. we use minimal effort :P.
Either way, I think I could go with something like ```SimpleCoherence``` which is simple in the sense that it resembles coherence but it has no transitivity requirement. ```SimplifiedCoherence``` on the other hand sounds like it is computed from a true coherence by simplifying it, which in turn would suggest something like a transitive reduction (i.e. maximal simplification).

Reply:
There is still a reference to `woGraph()` in a comment of `Reasoner`